cms/utils/__init__.py

```diff
@@ -16,15 +21,42 @@ def get_current_site():
     return Site.objects.get_current()
 
 
+@lru_cache(512)
+def get_object_language(request):
+    try:
+        resolved = resolve(request.path)
+    except Resolver404:
+        return None
+
+    args = resolved.args
+    if len(args) != 2 or not all([a.isnumeric() for a in args]):
+        return None
+
+    content_type_id, object_id = args
+
+    try:
+        content_type = ContentType.objects.get_for_id(content_type_id)
+        content_type_obj = content_type.get_object_for_this_type(pk=object_id)
+        language = content_type_obj.language
+    except (ContentType.DoesNotExist, AttributeError):
+        return None
+
+    return language
+
+
 def get_language_from_request(request, current_page=None):
     """
     Return the most obvious language according the request
     """
     language = None
     if hasattr(request, 'POST'):
         language = request.POST.get('language', None)
-    if hasattr(request, 'GET') and not language:
+    elif hasattr(request, 'GET'):
         language = request.GET.get('language', None)
+
+    if not language:
```

Comment:
Can you add a comment why we need to get the language here, maybe enforcing why you haven't used the fallback language. There arn't enough comments in the core. They don't have to be every line but defo anytime we need to do something out of the ordinary or not obvious why.